python/tskit/trees.py

```diff
@@ -4536,6 +4546,20 @@ def edge(self, id_):
             metadata_decoder=self.table_metadata_schemas.edge.decode_row,
         )
 
+    def _tree_node_edges(self):
+        """
+        Return a generator over the trees in the tree sequence, yielding a numpy array
+        that maps the node IDs in the tree to the ID of the edge above them.
+        Currently this is a private, non-optimised Python implementation.
+        """
+        node_edges = np.full(self.num_nodes, NULL, dtype=np.int32)
+        for _, edges_out, edges_in in self.edge_diffs():
```

Comment:
I have no idea why this one is marked as partial

Reply:
This one also never gets to `StopIteration` as the caller stops early.

Reply:
I don't quite get why. If I do `zip(ts.trees(), ts.edge_diffs())` does it not reach `StopIteration` simultaneously on both iterators?

Reply:
Ah, no. Zip will always take the shortest iterator, so won't check the rest of the iters.

Reply:
Try reversing the order.

Reply:
No, just on one of them

Reply:
Or itertools.zip_longest should exhaust bother iters.

Reply:
Just FYI, Python 3.10 adds a boolean `strict` keyword arg to `zip()` which when `True` will raise a `ValueError` if the iterators don't all have the same length (consuming an equal number of elements from each iterator as a side effect).
https://www.python.org/dev/peps/pep-0618/